packages/flutter/lib/src/material/stretch_overscroll_effect.dart

```diff
@@ -0,0 +1,176 @@
+  ///
+  /// The value should be between -1.0 and 1.0 inclusive.
+  /// Positive values apply a pull from top to bottom,
+  /// while negative values pull from bottom to top.
+  ///
+  /// {@tool snippet}
+  /// This example shows how to set the vertical stretch strength to pull bottom.
+  ///
+  /// ```dart
+  /// StretchOverscrollEffect(
+  ///   stretchStrengthY: 0.5,
+  ///   child: Text("Hello, World!"),
+  /// );
+  /// ```
+  /// {@end-tool}
+  final double stretchStrengthY;
+
+  /// The child widget that receives the stretching overscroll effect.
+  final Widget child;
+
+  @override
+  State<StretchOverscrollEffect> createState() => _StretchOverscrollEffectState();
+}
+
+class _StretchOverscrollEffectState extends State<StretchOverscrollEffect> {
+  ui.FragmentShader? _fragmentShader;
+
+  /// The maximum scale multiplier applied during a stretch effect.
+  static const double maxStretchIntensity = 1.0;
+
+  /// The strength of the interpolation used for smoothing the effect.
+  static const double interpolationStrength = 0.7;
+
+  @override
+  void dispose() {
+    _fragmentShader?.dispose();
+    super.dispose();
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    if (!_StretchOverscrollEffectShader._initCalled) {
+      _StretchOverscrollEffectShader.initializeShader();
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final bool isShaderNeeded =
+        widget.stretchStrengthX.abs() > precisionErrorTolerance ||
+        widget.stretchStrengthY.abs() > precisionErrorTolerance;
+
+    final ui.ImageFilter imageFilter;
+
+    if (_StretchOverscrollEffectShader._initialized) {
+      _fragmentShader?.dispose();
+      _fragmentShader = _StretchOverscrollEffectShader._program!.fragmentShader();
+      _fragmentShader!.setFloat(2, maxStretchIntensity);
+      _fragmentShader!.setFloat(3, widget.stretchStrengthX);
+      _fragmentShader!.setFloat(4, widget.stretchStrengthY);
+      _fragmentShader!.setFloat(5, interpolationStrength);
+
+      imageFilter = ui.ImageFilter.shader(_fragmentShader!);
+    } else {
+      _fragmentShader?.dispose();
+      _fragmentShader = null;
+
+      imageFilter = ui.ImageFilter.matrix(Matrix4.identity().storage);
+    }
+
+    return ImageFiltered(
+      imageFilter: imageFilter,
+      enabled: isShaderNeeded,
+      // A nearly-transparent pixels is used to ensure the shader gets applied,
+      // even when the child is visually transparent or has no paint operations.
+      child: CustomPaint(
+        painter: isShaderNeeded ? _StretchOverscrollEffectPainter() : null,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+/// CustomPainter that draws nearly transparent pixels at the four corners.
+///
+/// This ensures the fragment shader covers the entire canvas by forcing
+/// painting operations on all edges, preventing shader optimization skips.
+class _StretchOverscrollEffectPainter extends CustomPainter {
```

Comment:
For my information, what would happen if we don't have this? In what situation will shader skips?